Use local Google Fonts

drmenzelit opened the pull request:
Pull Request for Issue #14 #15  .

### Summary of Changes
Added parameter(s) for GoogleFonts. 
Example fonts installed.
Removed Fira Sans

### Testing Instructions
Open template, you will see two new parameters
Google Font yes / no
Google Font Name as select field

![Screenshot_2020-08-12 Templates Edit Style - Joomla Cassiopeia Enhancement - Administration](https://user-images.githubusercontent.com/9153168/90011866-a2102080-dca2-11ea-87a0-a0f042d9c4e7.png)

### Expected result
If "no" is selected the template will use the font definitions from BS4 (at the moment)
If "yes" you can select a font or font combination. The selected font will be loaded with a specific css file.

![Screenshot_2020-08-12 Home](https://user-images.githubusercontent.com/9153168/90011894-ac321f00-dca2-11ea-949b-f6c6f6f79ba2.png)
![Screenshot_2020-08-12 Home-Josefin](https://user-images.githubusercontent.com/9153168/90011898-afc5a600-dca2-11ea-83ea-0596290a8713.png)
![Screenshot_2020-08-12 Home-nofont](https://user-images.githubusercontent.com/9153168/90011901-b3592d00-dca2-11ea-9d7b-6e9f750d109c.png)

### Documentation Changes Required

We need to explain how to install and use other fonts

#### Update 2020-08-21 by @richard67 

There are 2 use cases.

Case 1: CMS contibutors add new fonts to the sources so they are shipped with the CMS.

This should not really happen often since we should at the end ship only the fonts which are really used by the CMS, maybe a choice of 2 which look nice, but not a lot.

Upload a **s**css file for each font or combination of fonts and corresponding woff and woff2 files to folder `/build/media_source/fonts` and then run `npm ci`. That's it.

Compiled css and copied font files will be availabe in folder `/media/fonts` after that.

In the source code only what has changed or been added in folder `/build/media_source/fonts` has to be checked in.

Case 2: Users add new fonts to an installlled J4.

Upload a css file for each font or combination of fonts and corresponding woff and woff2 files to folder `/media/fonts`. Optionally you can also upload the minified css file in addition. That's it.

In both cases, after you have done the above steps (either case 1 or case 2), you can select in backend in Cassiopeia's template style settings the new fonts by the base name of the css files in the `/media/fonts` folder. Minified css files will be ignored.

drmenzelit committed:
Use local Google Fonts

Added parameter(s) for GoogleFonts. Example fonts installed.

drmenzelit committed:
Code improvement

renamed scss files, changed if to switch

richard67 commented:
@Fedik Hello, maybe you are surprised being pinged from this repository. We are preparing some extensions of the Cassiopeia template, and one of them is providing local Google fonts and preload them. Now my question: The web asset manager which you have added to the core, can it also load fonts? Or do we sill have to use `$this->getPreloadManager()->preload(...)` like it is done here? [https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Document/Document.php#L844](https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Document/Document.php#L844).

Fedik commented:
You still have to use `$this->getPreloadManager()`

WebAsset just provide an URI to the resource, if you example have an asset for fonts you can: 
```php
$wa  = $this->getWebAssetManager();
$wa->useStyle('template.cassiopeia.googlefont');

$this->getPreloadManager()
  ->preload($wa->getAsset('style','template.cassiopeia.googlefont')->getUri(), ['as' => 'style']);
```
Btw, I forgot to make this change in the joomla core :smiley: 
(So people can override a 'template.cassiopeia.googlefont' with plugin, without editing index.php)

And note about the changes, you better make a local `font.css` with defined local fonts, and preload this css file, instead of preloading the font files. In this way browser will pick up best font file from the font.css, without loading all of them.

Fedik commented:
This should work:
```php
switch( $this->params->get('googleFontName'))
{
  case "josefin-sans":
    $wa->useStyle('font.josefin');
    $this->getPreloadManager()->preload($wa->getAsset('style','font.josefin')->getUri(), ['as' => 'style']);
    break;

  case "montserrat":
    $wa->useStyle('font.montserrat');
    $this->getPreloadManager()->preload($wa->getAsset('style','font.montserrat')->getUri(), ['as' => 'style']);
    break;
}

```

drmenzelit committed:
Change preload

drmenzelit commented:
@Fedik thank you for your help. I made the changes as you suggested. The css file is being loaded so
<link href="/cassiopeia/templates/cassiopeia/css/font--montserrat.min.css?db8b2ee4897b7a67b66f3b60f9967527" rel="stylesheet" />
Is it correct or should we expect a rel=preload?

Fedik commented:
Yes, it is correct.
If you inspect the headers in the dev console (network tab), you should see something like this (but with your font file):

![Screenshot_2020-08-13_10-51-29](https://user-images.githubusercontent.com/1568198/90108525-1e623c80-dd53-11ea-98e3-65c3595050ec.png)

hans2103 commented:
@Fedik I am trying to add the woff and woff2 files to cassiopeia/joomla.asset.json
I get a notice in PHPStorm

> Value should be one of: "style", "script", "preset"

<img width="421" alt="Schermafbeelding 2020-08-13 om 10 52 11" src="https://user-images.githubusercontent.com/639822/90114652-4ea9db00-dd53-11ea-8ac3-c70a7f511343.png">


I would like to be able to add `type: "font",`
I would like to see:
```
<link href="/templates/cassiopeia/font/josefin-sans-v16-latin-600.woff" rel="preload" as="font" type="font/woff2" crossorigin="anonymous">
<link href="/templates/cassiopeia/font/josefin-sans-v16-latin-600.woff2" rel="preload" as="font" type="font/woff2" crossorigin="anonymous">
```

hans2103 commented:
@Fedik sorry.. had to read all of your comments first

> And note about the changes, you better make a local `font.css` with defined local fonts, and preload this css file, instead of preloading the font files. In this way browser will pick up best font file from the font.css, without loading all of them.

richard67 commented:
@Fedik Thanks a lot, you've helped much.

N6REJ committed:
simplify font call

N6REJ committed:
Merge origin/4.0-dev-google-fonts-enhancements into switch

Conflicts:
	templates/cassiopeia/index.php

N6REJ committed:
Merge origin/4.0-dev-google-fonts-enhancements into switch

Conflicts:
	templates/cassiopeia/index.php

hans2103 committed:
:truck: rename josefin into josefin-sans

hans2103 committed:
Merge branch 'cassiopeia/4.0-dev-google-fonts-enhancements' into feature-cassiopeia/4.0-dev-google-fonts-enhancements

richard67 committed:
Apply suggestions from code review

richard67 committed:
Merge pull request #21 from N6REJ/switch

Replace switch

richard67 committed:
Merge pull request #22 from hans2103/4.0-dev-google-fonts-enhancements

4.0 dev google fonts enhancements

richard67 commented on templates/cassiopeia/scss/_variables.scss:
```diff
@@ -77,9 +77,9 @@ $theme-colors: (
 
 // Fonts
-$font-family-base:                   "Fira Sans", Arial, sans-serif;
+//$font-family-base:                   "Fira Sans", Arial, sans-serif;
```
Is it right that this is commented out? If so, it should be removed instead. Or is it just a remainder from testing?

N6REJ replied:
// is a rem

richard67 replied:
@N6REJ Bear I know what a comment is. I ask why it is commented out. Man!!! We shall not have commented out code in our code.

drmenzelit replied:
The line can be removed. That is a override of the font-family coming from BS4.

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-google-fonts-enhancements

richard67 commented:
Recent changes due to merged PR's may have broken it. At least I don't see it working here. Will continue to investigate.

richard67 commented:
Problem is that `$wa->getAsset('style', $googleFontName)->getUri()` returns empty.

richard67 commented:
The problem is that the compiled css is missing. It seems npm doesn't compile the `./templates/cassiopeia/scss/font--josefin-sans.scss` and `./templates/cassiopeia/scss/font--montserrat.scss` files.

richard67 commented:
@drmenzelit @hans2103 Ha that was it! The fonts scss files have to be added to the build script `./build/build-modules-js/compilecss.es6.js`:

```
...
        files = [
          `${RootPath}/templates/cassiopeia/scss/offline.scss`,
          `${RootPath}/templates/cassiopeia/scss/font--josefin-sans.scss`,
          `${RootPath}/templates/cassiopeia/scss/font--montserrat.scss`,
          `${RootPath}/templates/cassiopeia/scss/template.scss`,
          `${RootPath}/templates/cassiopeia/scss/template-rtl.scss`,
...
```

The question is if that it really open for adding fonts later in overrides.

Wouldn't it be better we add css files for the fonts to the css folder of the template instead of scss files which need to be compiled?

richard67 commented:
If I could chose, I would make css and put them into a folder "templates/cassiopeia/css/fonts" and remove the "font--" prefix from their name.

Only thing what I don't know is if it needs the step to compile from scss to css for some reason. If that is the case, then I would put the scss files into a folder "templates/cassiopeia/scss/fonts" and also remove the "font--" prefix from their name, and let the script use that folder in the "folders" array and not the single files in the "files" array.

richard67 committed:
[Google Fonts] Move the fonts scss files to a sub-folder and add them to the css compile script (#23)

* Rename templates/cassiopeia/scss/font--josefin-sans.scss to templates/cassiopeia/scss/fonts/josefin-sans.scss

* Rename templates/cassiopeia/scss/font--montserrat.scss to templates/cassiopeia/scss/fonts/montserrat.scss

* Fix PHP and SCSS code style

* More PHP code style

* Finish the sub folder move

* Add fonts scss to the compile script

richard67 commented:
I've solved it so far. The only question which remains is the one in my review comment above:

[https://github.com/joomla/cassiopeia/pull/18#discussion_r470960446](https://github.com/joomla/cassiopeia/pull/18#discussion_r470960446)

@drmenzelit Could you check and comment?

N6REJ commented:
> @drmenzelit @hans2103 Ha that was it! The fonts scss files have to be added to the build script `./build/build-modules-js/compilecss.es6.js`:
> 
This is not extensible.  Either use scss or css.  The nice thing about scss is it does allow for the base to be a $var which makes it a single point of change for end-users.
IF the .js can be told to look for 'fonts.scss' & use it if it exists then thats fine. ( pretty sure it can ) 
Then you'd have fonts.scss call the _variables to sort things out.
Ultimately a .css will be used so there doesn't HAVE to be a .scss as only dev's will have the tools to process it into a .css file anyway.
In the end you want 
/css
/fonts

in /css you want
template.css
fonts.css
custom.css

in /fonts you want only your .ttf & .woff/woff2 files.

richard67 commented:
@N6REJ 

> This is not extensible.

Nonsense. If one wants to extend it he or she can just put css files into the folder where now with my latest changes the css files which are compiled from the scss files go to.

Better you look on this branch like it is now with the last changes, and maybe check it out and run npm ci and test it before commenting.

richard67 committed:
Remove obsolete line

richard67 commented:
I think this is ready for testing now. Just checkout the branch of this PR, pull changes if necessary, and then run `npm ci`.

@ciar4n I'd be happy if you could have a quick look on it and let us know if it is good like it is now.

richard67 commented:
I think if this PR here is ok we shouldn't merge it here, we should instead made a PR for the CMS with the same branch, because this here is completely independent from other changes we will do.

N6REJ commented:
> @N6REJ
> 
> > This is not extensible.
> 
> Nonsense. If one wants to extend it he or she can just put css files into the folder where now with my latest changes the css files which are compiled from the scss files go to.
> 
> Better you look on this branch like it is now with the last changes, and maybe check it out and run npm ci and test it before commenting.

sorry, apparently my statement wasn't clear enough.. requiring the build script to be modified by the dev each time they want to change fonts ( as shown here ) is not extensible.
...
        files = [
          `${RootPath}/templates/cassiopeia/scss/offline.scss`,
          `${RootPath}/templates/cassiopeia/scss/font--josefin-sans.scss`,
          `${RootPath}/templates/cassiopeia/scss/font--montserrat.scss`,
          `${RootPath}/templates/cassiopeia/scss/template.scss`,
          `${RootPath}/templates/cassiopeia/scss/template-rtl.scss`,
...

richard67 commented:
It is as extendable as the whole template is. If you do it on the original template it will be lost when updating, and if you copy the template you can do in the copy what you want.

But I can change the build script to use the folder, then it is not necessary to edit it every time.

I‘d just like to have some feedback from @ciar4n if that’s a way to go.

richard67 commented:
@N6REJ Please test #25 .

richard67 committed:
Use folder and not files for the fonts scss in build script for better extensibility (#25)

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-google-fonts-enhancements

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-google-fonts-enhancements

richard67 commented:
Maybe we should extend this by ~~an option to switch on~~ bold weight for the body font? (Thanks @ciar4n for that idea).

**Update:** Should not be an option, I missunderstood. Bold font should always be loaded, so we don't rely on a cached version of the font or the fallback (secondary font) is used when having bold text.

SharkyKZ commented:
Maybe the CMS shouldn't be used as garbage bin.

richard67 commented:
> Maybe the CMS shouldn't be used as garbage bin.

@SharkyKZ Could you enlight us what you mean? What is the garbage? The local fonts? We need that to be safe with GDPR. Or this code here? Feel free to comment it so we can improve.

Our plan is that we will make a **_regular_** PR for the CMS for this work piece, which then will go through the usual reviews and tests.

We do **_not_** plan to merge allwork in this repository together here and then make one mega monster PR which will be merged without the required reviews and tests.

SharkyKZ commented:
Nothing in this PR makes sense to me. First, the font chosen by original Cassiopeia designers has been replaced by two completely random, unfitting fonts. They are randomly dropped in template folder with no hint as to where they're sourced from or how they should be maintained. Preferably they should be pulled in through a dependency manager instead.

Then the `Documentation Changes Required` section implies all of this is done to add custom font support and those random fonts are just examples. But having to copy the template, modify its files and upload own files hardly counts as supporting anything. The same can be achieved without this PR.

N6REJ commented:
@SharkyKZ The fonts are clearly google fonts.
My personal thought is to have a "fonts" form field that pulls from the api the list of available fonts and either renders it as a list or as a live(?) list that shows the look of the font chosen.
then, when they've selected a font that is then dl'd, put into the fonts folder and the appropriate web component created.
Older fonts could be removed at this time if thats what we want to do.
The purpose of having a form field is so that other font sources could be used not just google.
Literally EVERY template in existence has google fonts as part of the template now days.
In my personal template I took it a step farther and made it so that the fonts were selectable by element but I was told by a designer we all know that it's kind of overwhelming in the repeatable form field way, I presented it.
Presenting a list is relatively simple.

N6REJ commented:
@richard67 perhaps we need to look @ what fonts work together from a design point of view?  Similar to contrast effects

chmst commented:
Work on the Design ist still in Progress

richard67 commented:
Changed back to draft so we don't accidently merge it.

richard67 commented:
@Fedik Sorry, but I need again your advise. I have 1 question: Can the same what we have here now be also achieved without having to pre-define the available font css files in file `templates/cassiopeia/joomla.asset.json`?

I'd like to simplify adding new fonts.

Currently one would have to edit the `templates/cassiopeia/templateDetails.xml` for adding the font to the options of the select drop down, and also edit the `templates/cassiopeia/joomla.asset.json` to define the assets.

I'd like to avoid both by getting the available css files for the options from the file system, and loading the right css file (which then defines the fonts) based on the elected option. If someone puts new fonts and css file, I would like it to work out of the box.

Is that possible? And if yes, is there anything else speaking against it, e.g. security reasons?

richard67 committed:
Merge remote-tracking branch 'joomla-cms/4.0-dev' into 4.0-dev-google-fonts-enhancements

richard67 committed:
Merge remote-tracking branch 'joomla-cms/4.0-dev' into 4.0-dev-google-fonts-enhancements

Fedik commented:
`joomla.asset.json` just for static stuff.

You can add any asset dynamically, eg for css:
```php
$wa->registerStyle('font.blabla', 'fonts-blabla.css');
$wa->useStyle('font.blabla');

// or one row
$wa->registerAndUseStyle('font.blabla', 'fonts-blabla.css');
```

You can use what is better for you, there is no restriction.

richard67 commented:
@Fedik And again thanks for your advise. In fact I already tried what you suggested, but it didn't work. Maybe I made some mistake. I'll check and if possible fix it with help of the doc https://docs.joomla.org/J4.x:Web_Assets (thank you also for that doc).

richard67 commented:
@Fedik Ha, first question comes soon: In a template "X", can I register and use also css files from another template "Y"? I'll try that soon, but maybe you are faster than my experiments ;-)

Fedik commented:
> In a template "X", can I register and use also css files from another template "Y"?

Yes, you can register any files, as long as you define correct path. However it will not take in count overrides that exist for other template. So in general it not very good idea.

It resolve files with HtmlHelper,

```php
$wa->registerAndUseStyle('font.blabla', 'fonts-blabla.css');
// old would be
HTMLHelper::_('stylesheet', 'fonts-blabla.css' );
```

With this it will look only active template (and overides), for use this file from another template you have to set "full relative path" to the file.

Better to avoid it, it may end up not much nice. 
If need a shared asset then better put it in `/media`

richard67 commented:
@Fedik Yeah, I meanwhile have it working for the same template. Will make PR against this one here soon. It will not need anymore to edit the json and the xml for adding a new font.

I had the idea to do it like we do it in modules with the chrome style select, that you can select styles from all templates.

But I think you are right, for the fonts css it is maybe not a good idea, those should be limited to the same template, or maybe in addition get some from the system template.

Thanks a lot, Fedir.

Fedik commented:
If your idea is to have a shared fonts, then maybe need to think about `/media/fonts` folder, where User can upload the folder with fonts and its css:
```
/media/fonts/foo/foo-fonts.css
/media/fonts/foo/foo-fonts.xx
/media/fonts/foo/foo-fonts.xx

/media/fonts/bar/bar-fonts.css
/media/fonts/bar/bar-fonts.xx
/media/fonts/bar/bar-fonts.xx
```

Fedik commented:
and in template make a config field, for the path to `foo/foo-fonts.css`
and load it in template with `$wa->registerAndUseStyle()`

so User do not need to edit anything, except upload files, and set template params

richard67 commented:
> If your idea is to have a shared fonts, then maybe need to think about `/media/fonts` folder, where User can upload the folder with fonts and its css:
> 
> ```
> /media/fonts/foo/foo-fonts.css
> /media/fonts/foo/foo-fonts.xx
> /media/fonts/foo/foo-fonts.xx
> 
> /media/fonts/bar/bar-fonts.css
> /media/fonts/bar/bar-fonts.xx
> /media/fonts/bar/bar-fonts.xx
> ```

Good idea, I like it.

@chmst @drmenzelit @hans2103 Has that been discussed in the team before? I ask because I was not in the team from beginning on and so might not know all history of decisions.

richard67 commented:
> and in template make a config field, for the path to `foo/foo-fonts.css`
> and load it in template with `$wa->registerAndUseStyle()`
> 
> so User do not need to edit anything, except upload files, and set template params

Exactly that's what I'm currently working on.

Fedik commented:
just an idea, it need to test, I am not sure how useful it can be :wink:

richard67 commented:
It works here already with fonts in the template's folder structure.

To move that to media folder we have to discuss in the team, I think. Can't decide that alone.

Fedik commented:
You can just make in your way, no one will chage it back, people are lazy to do changes if it works already :smile: 
Just joking :wink:

richard67 commented:
Just a reminder for all readers: This PR is just for making development visible, it is not (yet) what we want to deliver to the CMS. E.g. the fonts we use here in this PR are just for testing. Maybe we should have used Comic-Sans to make it more clear ;-) And as discussions above show, development is still ongoing.

richard67 commented:
@Fedik Sure you meant the "media" folder? Or did you mean "images" folder where people can upload stuff with media manager? The "media" folder doesn't exist in sources, it is built by the build scripts. And "images/fonts", well, that would work nice with the media manager, but on the filesystem it look weird, having a folder "fonts" in a folder with name "images" ;-) But maybe we should do it that way, use "images/fonts", and if people want they can (but not need to) organise the fonts in sub-folders?

Fedik commented:
Yeah, I meant `/media` where all assets live, with FTP upload.

Technically user can upload it anywhere, as long as he/she set correct path to font css.
I would not place "core fonts" in /images because it weird at least :smile: 


btw, we probably should rename this /images some day it really confusing, /images/documets/bla.pdf, etc

richard67 commented:
Some enhancements regarding fonts location and having to edit files or not for adding new fonts see PR #39 .

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-google-fonts-enhancements

richard67 committed:
[Google Fonts] Local fonts enhancements (#39)

- Move the fonts abd their scss to `build/media_source/fonts`
- Modify the the css build script so it copies the fonts and the compiled css to `media/fonts`.
- Use a `filelist` field for selecting the fonts => No need to edit `templateDetails.xml` anymore for adding new fonts.
- Dynamically register the asset so no need to edit `joomla.asset.json` anymore for adding new fonts.

richard67 commented on templates/cassiopeia/index.php:
```diff
@@ -18,6 +18,9 @@
 $wa  = $this->getWebAssetManager();
 
+// Template path
+$templatePath = Uri::root() . 'templates/'.$this->template;
```
@drmenzelit Do you know if we need this variable? It was added with the first version of the Google Fonts here on this branch. I see it nowhere used now in template.php.

drmenzelit committed:
Added bolds and italics of Josefin Sans and Fira Sans

Removed Montserrat / Work Sans combination and added Fira Sans as is in Cassiopeia v1.

richard67 commented:
When I open the template style options after a new installation, I see the Google Fonts option is already switched on, and fira-sans is selected. But the frontpage doesn't use that unless I save the template style options, without having changed anything else.

After that first saving, everything works well.

I think it is because the default value of the switcher as defined in the form XML is not equal to what you get in the PHP code when that configuration value is not set.

I'll see if I can fix it, I have an idea.

brianteeman commented:
@richard67 that means its not being defined in the installation sql

richard67 commented:
@brianteeman Should it be, or could we use `if ($this->params->get('googleFont', true))` instead of `if ($this->params->get('googleFont'))` here? https://github.com/joomla/cassiopeia/pull/18/files#diff-040c75e93d662244ae546847f23bd32cR41 .. that would work, too, but maybe it's not a good idea.

brianteeman commented:
It should be in the sql

richard67 commented:
Hmm installation SQL is easy, but making an update SQL to update older 4.0 Betas would maybe be too risky, with search and replace in the options. I have to think about that.

richard67 committed:
Init googleFont parameter in installation SQL

richard67 committed:
Init googleFontName parameter in installation SQL

N6REJ commented:
on a fresh install the default used in the xml is what is set in the db as a param.  After the 1st save then it changes to whatever the values in the template are.  There's no need to have a value preset in a sql file.
in order to prove this remove all params from the db and reload the template.  It will populate based off the defaults.  Ran into this many many times building my template a couple of years ago.

richard67 commented:
@N6REJ Sorry Troy, but it seems you did not get the problem. When you make a new installation and you do not do anything in backend and you go to the site, the template checks the result of `$this->params->get('googleFont')`. And because there is nothing in db and no default given, it returns false (or zero). The frontent template doesn't know anything about the default values of the templates styles form's fields in the backend.

If you now go to backend and check the settings without saving, you will see that the google fonts are swiched on and a font is selected, which does not fit to what you get in the frontent,

There are 2 ways out: 1. Init in db, and 2. using `$this->params->get('googleFont', true)`. But either of these ways means we define the default at 2 places, 1. in the form xml for the field, and 2. in db or the function call.

We had this problem at other places, too, and one of the 2 ways or both have been used to solve it, and we still have it at other places where it was not really solved. I think that's a weak point in design that we define the default values in a way which is not available if not using the form.

brianteeman commented:
If as a matter of best practice we get in the habit of always writing the defaults in the xml it will resolve a lot of problems as this "issue" keeps coming up.

Personally I would not care about beta to beta upgrades for this

richard67 commented:
@brianteeman 

> If as a matter of best practice we get in the habit of always writing the defaults in the xml it will resolve a lot of problems as this "issue" keeps coming up.

Really "xml"? Or should it have been "sql" and the spell checker "corrected" it?

> Personally I would not care about beta to beta upgrades for this

Me neither. We have broken it anyway with the child templates ~~, so it's just one more sentence to add to the release notes of the next beta about what to do when updating a previous beta~~ .

brianteeman commented:
not enough coffee yet. it was of course sql

richard67 commented:
~~Regarding the update sql I was silly: No need for a dangerous `replace` in the update statement. We can simply set the complete `params` value for the Cassiopeia template style in the same way as we do it for new installations. So at the end when this and other changes are ready, I can make an update SQL to fix params of those template styles.~~

Update: Forget it, we can't do that. One may alrerady have changed some of the params and will complain if that is reset with an update.

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-google-fonts-enhancements

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-google-fonts-enhancements

richard67 committed:
Remove obsolete fonts folder from template

richard67 commented:
Am merging into the development branch for now. For further improvements please make pull requests against the [development](https://github.com/joomla/cassiopeia/tree/development) branch.
